pkg/devfile/generator/generators.go

```diff
@@ -0,0 +1,327 @@
+package generator
+
+import (
+	"fmt"
+
+	buildv1 "github.com/openshift/api/build/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	extensionsv1 "k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	v1 "github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+
+	"github.com/devfile/library/pkg/devfile/parser"
+)
+
+const (
+	// DevfileSourceVolumeMount is the directory to mount the volume in the container
+	DevfileSourceVolumeMount = "/projects"
+
+	// EnvProjectsRoot is the env defined for project mount in a component container when component's mountSources=true
+	EnvProjectsRoot = "PROJECTS_ROOT"
+
+	// EnvProjectsSrc is the env defined for path to the project source in a component container
+	EnvProjectsSrc = "PROJECT_SOURCE"
+
+	deploymentKind       = "Deployment"
+	deploymentAPIVersion = "apps/v1"
+)
+
+// GetObjectMeta creates a common object meta
+func GetObjectMeta(name, namespace string, labels, annotations map[string]string) metav1.ObjectMeta {
+
+	objectMeta := metav1.ObjectMeta{
+		Name:        name,
+		Namespace:   namespace,
+		Labels:      labels,
+		Annotations: annotations,
+	}
+
+	return objectMeta
+}
+
+// GetContainers iterates through the components in the devfile and returns a slice of the corresponding containers
+func GetContainers(devfileObj parser.DevfileObj) ([]corev1.Container, error) {
+	var containers []corev1.Container
+	for _, comp := range devfileObj.Data.GetDevfileContainerComponents() {
+		envVars := convertEnvs(comp.Container.Env)
+		resourceReqs := getResourceReqs(comp)
+		ports := convertPorts(comp.Container.Endpoints)
+		containerParams := containerParams{
+			Name:         comp.Name,
+			Image:        comp.Container.Image,
+			IsPrivileged: false,
+			Command:      comp.Container.Command,
+			Args:         comp.Container.Args,
+			EnvVars:      envVars,
+			ResourceReqs: resourceReqs,
+			Ports:        ports,
+		}
+		container := getContainer(containerParams)
+
+		// If `mountSources: true` was set, add an empty dir volume to the container to sync the source to
```

Comment:
Are we sure that all projects needs this? emptyDir volume?

Reply:
This is just a comment, we are not doing any empty volume mounting. If mountSources is true, we just set `PROJECTS_ROOT` & `PROJECT_SOURCE` for the container.

I'll go through the comments to make sure they're updated and clean.